README.md

# Flutter Cache Image

[![pub package](https://img.shields.io/pub/v/cache_image.svg)](https://pub.dartlang.org/packages/cache_image)

A Flutter plugin to load and cache network or firebase storage images with a retry mechanism if the download fails.
 
This package supports the download of images from a standard *network path* and from a *firebase storage gs path*. 

Images are stored in the temporary directory of the app.

## Usage

To use this plugin, add firebase_storage as a dependency in your pubspec.yaml file.

```
dependencies:
  cache_image: "^1.0.5"

```

Import cache_image in a dart file:
```
import 'package:cache_image/cache_image.dart';
```

To support firebase storage download the generated google-services.json file and place it inside android/app. Next, modify the android/build.gradle file and the android/app/build.gradle file to add the Google services plugin as described by the Firebase assistant. 

## How to use

Cache Image can be used with any widget that support an ImageProvider.

``` dart
Image(
    fit: BoxFit.cover,
    image: CacheImage('gs://your-project.appspot.com/image.png'),
),
Image(
    fit: BoxFit.cover,
    image:  CacheImage('https://hd.tudocdn.net/874944?w=646&h=284', duration: Duration(seconds: 2), durationExpiration: Duration(seconds: 10)),
),
FadeInImage(
    fit: BoxFit.cover,
    placeholder: AssetImage('assets/placeholder.png'),
    image: CacheImage('gs://your-project.appspot.com/image.jpg')
)
 ```

See the `example` directory for a complete sample app using Cache Image.

CHANGELOG.md
## 1.0.5

* Bug fix

## 1.0.4

* Doc updated
* Retry mechanism improved

## 1.0.3

* Code format

## 1.0.2

* Pub badge
* Package description update

## 1.0.1

* Doc updated
* Example updated

## 1.0.0

* Stable release
* Update flutter minimum version
* Support to auto-retry if download fails


## 0.0.3

* Dependencies updated
* Bug fix

## 0.0.2

* Bug fix temp dir

## 0.0.1

* Initial release, support to firebase and standard network 

test/cache_image_test.dart
import 'package:cache_image/cache_image.dart';

void main() {}


lib/resource.dart
library cache_image;

import 'dart:io';
import 'dart:typed_data';
import 'package:firebase_storage/firebase_storage.dart';
import 'package:flutter/foundation.dart';
import 'package:path_provider/path_provider.dart';

class Resource {
  final String uri;
  final Duration duration;
  final double durationMultiplier;
  final Duration durationExpiration;

  Uri _temp;
  Uri _local;
  Uri _remote;
  Duration _retry;

  Resource(
      this.uri, this.duration, this.durationMultiplier, this.durationExpiration)
      : assert(uri != null),
        _retry = duration;

  Uri get remote => _remote;
  Uri get temp => _temp;
  Uri get local => _local;

  Uri _parse(String uri) {
    return Uri.parse(uri);
  }

  Future<Uri> _getTempDir() async {
    final Directory temp = await getTemporaryDirectory();
    return _parse(temp.path);
  }

  Future<Resource> init() async {
    _temp = await _getTempDir();
    _remote = _parse(uri);
    _local = _parse(_temp.path + _remote.path);
    return this;
  }

  Future<bool> checkFile() async {
    final File file = File(_local.path);
    if (file.existsSync() && file.lengthSync() > 0) {
      return true;
    }
    return false;
  }

  Future<Uint8List> getFile() async {
    final File file = File(_local.path);
    if (file.existsSync() && file.lengthSync() > 0) {
      return file.readAsBytesSync();
    }
    return null;
  }

  Future<Uint8List> storeFile() async {
    File file = await File(_local.path).create(recursive: true);
    // Check FireStorage scheme
    if (_remote.scheme == 'gs') {
      final StorageReference ref =
          FirebaseStorage.instance.ref().child(_remote.path);
      final dynamic url = await ref.getDownloadURL();
      _remote = Uri.parse(url);
    }
    // Download file with retry
    while (file.lengthSync() <= 0) {
      await Future.delayed(_retry).then((_) async {
        try {
          HttpClient httpClient = new HttpClient();
          final HttpClientRequest request = await httpClient.getUrl(_remote);
          final HttpClientResponse response = await request.close();
          final Uint8List bytes = await consolidateHttpClientResponseBytes(
              response,
              autoUncompress: false);
          file = await file.writeAsBytes(bytes);
        } catch (err) {
          _retry += _retry * this.durationMultiplier;
        }
      });
      // Check duration expiration
      if (_retry > this.durationExpiration) {
        break;
      }
    }
    return file.readAsBytesSync();
  }
}


lib/cache_image.dart
library cache_image;

import 'dart:typed_data';
import 'dart:ui';
import 'dart:async';
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:flutter/cupertino.dart';
import 'package:cache_image/resource.dart';

/*
 *  ImageCache for Flutter
 *
 *  Copyright (c) 2019 Oxequa - Alessio Pracchia
 *  Keep in touch https://www.linkedin.com/in/alessio-pracchia/
 *
 *  Released under MIT License.
 */

class CacheImage extends ImageProvider<CacheImage> {
  CacheImage(
    String url, {
    this.scale = 1.0,
    this.cache = true,
    this.duration = const Duration(seconds: 1),
    this.durationMultiplier = 1.5,
    this.durationExpiration = const Duration(seconds: 10),
  })  : assert(url != null),
        _resource =
            Resource(url, duration, durationMultiplier, durationExpiration);

  /// The scale to place in the [ImageInfo] object of the image.
  final double scale;

  /// Enable or disable image caching.
  final bool cache;

  /// Retry duration if download fails.
  final Duration duration;

  /// Retry duration multiplier.
  final double durationMultiplier;

  /// Retry duration expiration.
  final Duration durationExpiration;

  Resource _resource;

  Future<Codec> _fetchImage() async {
    Uint8List file;
    await _resource.init();
    final bool check = await _resource.checkFile();
    if (check) {
      file = await _resource.getFile();
    } else {
      file = await _resource.storeFile();
    }
    if (file.length > 0) {
      return PaintingBinding.instance.instantiateImageCodec(file);
    }
    return null;
  }

  @override
  Future<CacheImage> obtainKey(ImageConfiguration configuration) {
    return SynchronousFuture<CacheImage>(this);
  }

  @override
  ImageStreamCompleter load(CacheImage key, DecoderCallback decode) {
    return MultiFrameImageStreamCompleter(
        codec: key._fetchImage(),
        scale: key.scale,
        informationCollector: () sync* {
          yield DiagnosticsProperty<ImageProvider>(
              'Image provider: $this \n Image key: $key', this,
              style: DiagnosticsTreeStyle.errorProperty);
        });
  }
}


android/app/src/main/java/io/flutter/plugins/GeneratedPluginRegistrant.java
package io.flutter.plugins;

import io.flutter.plugin.common.PluginRegistry;
import io.flutter.plugins.firebase.core.FirebaseCorePlugin;
import io.flutter.plugins.firebase.storage.FirebaseStoragePlugin;
import io.flutter.plugins.pathprovider.PathProviderPlugin;

/**
 * Generated file. Do not edit.
 */
public final class GeneratedPluginRegistrant {
  public static void registerWith(PluginRegistry registry) {
    if (alreadyRegisteredWith(registry)) {
      return;
    }
    FirebaseCorePlugin.registerWith(registry.registrarFor("io.flutter.plugins.firebase.core.FirebaseCorePlugin"));
    FirebaseStoragePlugin.registerWith(registry.registrarFor("io.flutter.plugins.firebase.storage.FirebaseStoragePlugin"));
    PathProviderPlugin.registerWith(registry.registrarFor("io.flutter.plugins.pathprovider.PathProviderPlugin"));
  }

  private static boolean alreadyRegisteredWith(PluginRegistry registry) {
    final String key = GeneratedPluginRegistrant.class.getCanonicalName();
    if (registry.hasPlugin(key)) {
      return true;
    }
    registry.registrarFor(key);
    return false;
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:flutter/cupertino.dart';
import 'package:cache_image/cache_image.dart';
import 'package:path_provider/path_provider.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    SystemChrome.setPreferredOrientations([
      DeviceOrientation.portraitUp,
      DeviceOrientation.portraitDown
    ]);
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primaryColor: Colors.white,
      ),
      home: MyHomePage(title: 'Flutter Cache Image'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {

  void _refresh() {
    getTemporaryDirectory().then((dir){
      dir.delete(recursive: true);
    });
    setState(() {});
  }

  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        backgroundColor: CupertinoColors.white,
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title, style: TextStyle(color: CupertinoColors.black),),
      ),
      body: Column(
        children: <Widget>[
          Expanded(
            child: Stack(
              children: <Widget>[
                Container(
                  height: MediaQuery.of(context).size.height,
                  child: FadeInImage(
                    fit: BoxFit.cover,
                    placeholder: AssetImage('assets/placeholder.png'),
                    image: CacheImage('gs://testing-9ea12.appspot.com/bake-1706051_1920.jpg')
                  ),
                ),
                Positioned(
                    top:  0,
                    right: 0,
                    child: Container(
                      margin: EdgeInsets.all(16.0),
                      padding: EdgeInsets.only(top: 8.0, bottom: 8.0, right: 16.0, left: 16.0),
                      decoration: BoxDecoration(
                          color: Colors.white,
                          boxShadow: [
                            new BoxShadow(
                              color: Colors.black54,
                              blurRadius: 10.0,
                              spreadRadius: 1.0,
                              offset: new Offset(2.0, 1.0),
                            )
                          ],
                          borderRadius: BorderRadius.circular(16.0)
                      ),
                      child: Text("Firebase Storage"),
                    )
                ),
              ],
            ),
          ),
          Container(height: 4.0),
          Expanded(
            child: Stack(
              children: <Widget>[
                Container(
                  height: MediaQuery.of(context).size.height,
                  child: FadeInImage(
                    fit: BoxFit.cover,
                    placeholder: AssetImage('assets/placeholder.png'),
                    image: CacheImage('https://hd.tudocdn.net/874944?w=646&h=284', duration: Duration(seconds: 2), durationExpiration: Duration(seconds: 10))
                  ),
                ),
                Positioned(
                  top:  0,
                  right: 0,
                  child: Container(
                    margin: EdgeInsets.all(16.0),
                    padding: EdgeInsets.only(top: 8.0, bottom: 8.0, right: 16.0, left: 16.0),
                    decoration: BoxDecoration(
                        color: Colors.white,
                        boxShadow: [
                          new BoxShadow(
                            color: Colors.black54,
                            blurRadius: 10.0,
                            spreadRadius: 1.0,
                            offset: new Offset(2.0, 1.0),
                          )
                        ],
                        borderRadius: BorderRadius.circular(16.0)
                    ),
                    child: Text("Network"),
                  ),
                )
              ],
            ),
          )
        ],
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _refresh,
        backgroundColor: Theme.of(context).primaryColor,
        child: Icon(Icons.refresh, color: CupertinoColors.black),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
